src/transactions/TransactionFrame.cpp

```diff
@@ -471,36 +471,36 @@ TransactionFrame::processFeeSeqNum(AbstractLedgerTxn& ltx, int64_t baseFee)
 }
 
 void
-TransactionFrame::removeUsedOneTimeSignerKeys(
-    SignatureChecker& signatureChecker, AbstractLedgerTxn& ltx)
+TransactionFrame::removeOneTimeSignerFromAllSourceAccounts(
+    AbstractLedgerTxn& ltx) const
 {
-    for (auto const& usedAccount : signatureChecker.usedOneTimeSignerKeys())
+    auto header = ltx.loadHeader();
+    auto ledgerVersion = header.current().ledgerVersion;
+
+    if (ledgerVersion == 7)
     {
-        removeUsedOneTimeSignerKeys(ltx, usedAccount.first, usedAccount.second);
+        return;
     }
-}
 
-void
-TransactionFrame::removeUsedOneTimeSignerKeys(
-    AbstractLedgerTxn& ltx, AccountID const& accountID,
-    std::set<SignerKey> const& keys) const
-{
-    auto account = stellar::loadAccount(ltx, accountID);
-    if (!account)
+    std::unordered_set<AccountID> accounts{getSourceID()};
+    for (auto& op : mOperations)
     {
-        return; // probably account was removed due to merge operation
+        accounts.emplace(op->getSourceID());
     }
 
-    auto header = ltx.loadHeader();
-    auto changed = std::accumulate(
-        std::begin(keys), std::end(keys), false,
-        [&](bool r, const SignerKey& signerKey) {
-            return r || removeAccountSigner(header, account, signerKey);
-        });
-
-    if (changed)
+    auto signerKey = SignerKeyUtils::preAuthTxKey(*this);
+    for (auto& accountID : accounts)
     {
-        normalizeSigners(account);
+        auto account = stellar::loadAccount(ltx, accountID);
+        if (!account)
+        {
+            return; // probably account was removed due to merge operation
```

Comment:
This looks wrong, because it's possible for existing source accounts to follow missing source accounts. You should also add a test to catch this error.